### PR TITLE
[MNG-6535] Improve test coverage of UrlNormalizer.

### PR DESCRIPTION
**Pull request by josephw**
Refine the original PR (#199) to the test that improves coverage and give it a clearer description of its behaviour.

Although this isn't a major change, it does improve coverage (this case works, but isn't currently tested), and resolves the original ticket.

Following this checklist to help us incorporate your 
contribution quickly and easily:

 - [X] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MNG) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without 
       pulling in other changes.
 - [X] Each commit in the pull request should have a meaningful subject line and body.
 - [X] Format the pull request title like `[MNG-XXX] - Fixes bug in ApproximateQuantiles`,
       where you replace `MNG-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the 
       commit message.
 - [X] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [X] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.
 - [ ] You have run the [Core IT][core-its] successfully.

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under 
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [X] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [ ] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

[core-its]: https://maven.apache.org/core-its/core-it-suite/

### CODE REVIEW

File: maven-model-builder/src/test/java/org/apache/maven/model/path/DefaultUrlNormalizerTest.java [hunk 1]
```diff
@@ -82,4 +82,8 @@ public void testPreservationOfDoubleSlashes()
                       normalize( "[fetch=]http://server.org/[push=]ssh://server.org/" ) );
     }
 
+    public void testNormalizeInputWithNoParentDirectoriesToAscendTo()
+    {
+        assertEquals("a/../", normalize("a/../"));
+    }
```
Comment (rfscholte) [inline maven-model-builder/src/test/java/org/apache/maven/model/path/DefaultUrlNormalizerTest.java]: Maybe the implementation is wrong, but comparing this with Path this is the result:

        assertEquals( "a/../",  Paths.get("a/../").normalize().toString() ); // fails
        assertEquals( "",  Paths.get("a/../").normalize().toString() ); // succeeds

The Path behavior is also what I would expect.

Comment (michael-o) [reply maven-model-builder/src/test/java/org/apache/maven/model/path/DefaultUrlNormalizerTest.java]: @rfscholte So our impl is wrong?

Comment (josephw) [reply maven-model-builder/src/test/java/org/apache/maven/model/path/DefaultUrlNormalizerTest.java]: I'm not sure it's worth changing the behaviour; this class is used with absolute URLs (or paths), so this is very much an edge case.

However, I think it's (slightly) valuable to test that this case doesn't cause an error -- perhaps a different test name could make that clear?

Comment (michael-o) [reply maven-model-builder/src/test/java/org/apache/maven/model/path/DefaultUrlNormalizerTest.java]: If we normalize URLs, the test for relative paths must fail. URLs are absolute.

Comment (josephw) [reply maven-model-builder/src/test/java/org/apache/maven/model/path/DefaultUrlNormalizerTest.java]: Thinking again; there's no reason to retain the existing behaviour if it's wrong. The change to match `Path` is simple enough, and doesn't break the existing tests: I've pushed it.

Ideally, the interface would be clearer about its contract, but this is the least surprising behaviour.

File: maven-model-builder/src/test/java/org/apache/maven/model/path/DefaultUrlNormalizerTest.java [hunk 2]
```diff
@@ -82,8 +70,15 @@ public void testPreservationOfDoubleSlashes()
                       normalize( "[fetch=]http://server.org/[push=]ssh://server.org/" ) );
     }
 
-    public void testNormalizeInputWithNoParentDirectoriesToAscendTo()
+    @Test
+    public void relativeUriReferenceLeftUnaffectedWithNoParentDirectoryToAscendTo()
+    {
+        assertEquals( "/../", normalize("/../" ) );
```
Comment (michael-o) [inline maven-model-builder/src/test/java/org/apache/maven/model/path/DefaultUrlNormalizerTest.java]: Did you test the same with `Path`?

**Comment by josephw** [comment]
> Looks better now. Since you have changed (fixed) the impl, we need to turn the issue into a bug.

I've opened [MNG-6703](https://issues.apache.org/jira/browse/MNG-6703).

File: maven-model-builder/src/test/java/org/apache/maven/model/path/DefaultUrlNormalizerTest.java (hunk 2 above)
Comment (josephw) [reply maven-model-builder/src/test/java/org/apache/maven/model/path/DefaultUrlNormalizerTest.java]: `Path` and `URI` differ slightly here; I've gone with [RFC 3986's Remove Dot Segments](https://tools.ietf.org/html/rfc3986#section-5.2.4) and adjusted the code to remove them.

**Comment by michael-o** [comment]
PAsses for me locally, waiting for Jenkins..

**Comment by jira-importer** [comment]
Resolve #7314
